### PR TITLE
added support for RaspyRFM hardware

### PR DESCRIPTION
**Pull request by Phunkafizer**
Added support for RaspyRFM module:

http://www.seegel-systeme.de/produkt/raspyrfm-ii/

### CODE REVIEW

File: libs/pilight/hardware/raspyrfm.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+	Copyright (C) 2013 - 2014 Stefan Seegel
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.h]: Can you make this just 2017

File: libs/pilight/hardware/raspyrfm.c [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	div = gcd_a(rawlen, code);
+
+	for (i=0; i<rawlen; i++) {
+		code[i] /= div;
+		fifoByteCnt += code[i]; //count bits...
+	}
+	fifoByteCnt *= repeats;
+	fifoByteCnt = fifoByteCnt / 8 + 1; //mem we need for bit-FIFO
+
+	unsigned char *fifo = malloc(fifoByteCnt);
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.c]: Can you use the uppercase memory functions?

File: libs/pilight/hardware/raspyrfm.c [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	memset(fifo, 0x00, fifoByteCnt);
+
+	unsigned char mask = 0x80;
+	unsigned char *p = fifo;
+	int r;
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.c]: Can you initialize all variables when defining them.

File: libs/pilight/hardware/raspyrfm.c [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		if (wiringXSPISetup(rfmsettings->spi_ch, 250000) == -1) {
+			logprintf(LOG_ERR, "failed to open SPI channel: %d", rfmsettings->spi_ch);
+			return EXIT_FAILURE;
+		}
+
+		//check if rfm69 present
+		int i;
+		for (i=0; i<8; i++) {
+			writeReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i, 0x55);
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0x55)
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.c]: Can you use accolades everywhere, even when the if/else is just one line.

File: libs/pilight/hardware/raspyrfm.c [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+#define MODE_TX 3
+
+#define IRQFLAGS1_MODEREADY REGIRQFLAGS1 << 8 | 7
+
+#define IRQFLAGS2_FIFOFULL REGIRQFLAGS2 << 8 | 7
+#define IRQFLAGS2_FIFONOTEMPTY 6
+#define IRQFLAGS2_FIFOLEVEL 5
+#define IRQFLAGS2_PACKETSENT REGIRQFLAGS2 << 8 | 3
+
+
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.c]: Too many new lines.

File: libs/pilight/hardware/raspyrfm.c [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+#define WAITREGBITSET(x) while ((readReg(rfmsettings->spi_ch, (uint8_t) ((x) >> 8)) & (1<<((x) & 0x07))) == 0);
+#define WAITREGBITCLEAR(x) while ((readReg(rfmsettings->spi_ch, (uint8_t) ((x) >> 8)) & (1<<((x) & 0x07))) != 0);
+
+
```
Comment (CurlyMoo) [inline libs/pilight/hardware/raspyrfm.c]: Likewise.

**Comment by CurlyMoo** [comment]
I've added a few comments. Does this already work on the existing (433.92Mhz) protocols?

**Comment by Phunkafizer** [comment]
Yes, already wokring, tested with some intertechno sockets.
I pushed review. When okay I'll squash

**Comment by Phunkafizer** [comment]
Done "FREE"

**Comment by CurlyMoo** [comment]
Just curious. Is this device your making as well?

**Comment by CurlyMoo** [comment]
One other important thing. Can you publish under the MPLv2 license?

**Comment by Phunkafizer** [comment]
Sorry?

**Comment by Phunkafizer** [comment]
>Just curious. Is this device your making as well?
sorry...don't understand

**Comment by CurlyMoo** [comment]
Did you make the hardware as well?

**Comment by Phunkafizer** [comment]
>Did you make the hardware as well?
Yes I did... people like it maybe because they just have to plug it in...

**Comment by Phunkafizer** [comment]
>One other important thing. Can you publish under the MPLv2 license?
So changing the header of the files?

**Comment by CurlyMoo** [comment]
> Did you make the hardware as well?

I thought so. I will buy one soon to test it myself. 

> One other important thing. Can you publish under the MPLv2 license?

Yes

Would this work for other fifo hardware modules like the RFM43?

**Comment by Phunkafizer** [comment]
The RFM43 has very different registers, so there would be the need to write another module

**Comment by Phunkafizer** [comment]
>One other important thing. Can you publish under the MPLv2 license?

Do you have an example?

**Comment by CurlyMoo** [comment]
> Do you have an example?

The pilight core code in the rewrite branch.

**Comment by Phunkafizer** [comment]
Do I have to insert something in the cmake stuff so that the new modules can be selected?

**Comment by pilino1234** [comment]
You'll need to edit `CMakeLists.txt` and `CMakeExclude.txt` in `libs/pilight/hardware` and `CMakeConfig.txt` in the root pilight directory.

File: libs/pilight/hardware/CMakeLists.txt [hunk 7]
```diff
@@ -13,6 +13,12 @@ if(${MODULESPACK} MATCHES "OFF" OR WIN32)
 		list(REMOVE_ITEM ${PROJECT_NAME}_headers "${PROJECT_SOURCE_DIR}/433lirc.h")
 		list(REMOVE_ITEM ${PROJECT_NAME}_sources "${PROJECT_SOURCE_DIR}/IRgpio.c")
 		list(REMOVE_ITEM ${PROJECT_NAME}_headers "${PROJECT_SOURCE_DIR}/IRgpio.h")
+        list(REMOVE_ITEM ${PROJECT_NAME}_headers "${PROJECT_SOURCE_DIR}/raspyrfm.h")
+        list(REMOVE_ITEM ${PROJECT_NAME}_sources "${PROJECT_SOURCE_DIR}/raspyrfm.c")
+        list(REMOVE_ITEM ${PROJECT_NAME}_headers "${PROJECT_SOURCE_DIR}/433raspyrfm.h")
+        list(REMOVE_ITEM ${PROJECT_NAME}_sources "${PROJECT_SOURCE_DIR}/433raspyrfm.c")
+        list(REMOVE_ITEM ${PROJECT_NAME}_headers "${PROJECT_SOURCE_DIR}/868raspyrfm.h")
+        list(REMOVE_ITEM ${PROJECT_NAME}_sources "${PROJECT_SOURCE_DIR}/868raspyrfm.c")
```
Comment (pilino1234) [inline libs/pilight/hardware/CMakeLists.txt]: Use tabs for inline markup on these lines.

File: libs/pilight/hardware/raspyrfm.c [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+static uint8_t readReg(uint8_t spi_ch, uint8_t reg) {
+	uint8_t tmp[] = {reg, 0};
+	wiringXSPIDataRW(spi_ch, (unsigned char*) tmp, sizeof(tmp));
+	return tmp[1];
+}
+
+static int gcd(int a, int b) {
+	while (b != 0) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: Remove space after `while`

File: libs/pilight/hardware/raspyrfm.c [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	return EXIT_SUCCESS;
+}
+
+unsigned short raspyrfmHwInit(rfm_settings_t *rfmsettings) {
+	if(wiringXSupported() == 0) {
+		if(wiringXSetup() == -1) {
+			return EXIT_FAILURE;
+		}
+
+		if (wiringXSPISetup(rfmsettings->spi_ch, 250000) == -1) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: remove space after if

File: libs/pilight/hardware/raspyrfm.c [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+			logprintf(LOG_ERR, "failed to open SPI channel: %d", rfmsettings->spi_ch);
+			return EXIT_FAILURE;
+		}
+
+		//check if rfm69 present
+		int i;
+		for (i=0; i<8; i++) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: remove space after for

File: libs/pilight/hardware/raspyrfm.c [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0x55) {
+				break;
+			}
+			writeReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i, 0xAA);
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0xAA) {
+				break;
+			}
+		}
+
+		if (i < 8) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: remove space after if

etc. etc. throughout the file.

File: libs/pilight/hardware/raspyrfm.c [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+		//check if rfm69 present
+		int i;
+		for (i=0; i<8; i++) {
+			writeReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i, 0x55);
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0x55) {
+				break;
+			}
+			writeReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i, 0xAA);
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0xAA) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: remove space after if

File: libs/pilight/hardware/raspyrfm.c [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+		if (wiringXSPISetup(rfmsettings->spi_ch, 250000) == -1) {
+			logprintf(LOG_ERR, "failed to open SPI channel: %d", rfmsettings->spi_ch);
+			return EXIT_FAILURE;
+		}
+
+		//check if rfm69 present
+		int i;
+		for (i=0; i<8; i++) {
+			writeReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i, 0x55);
+			if (readReg(rfmsettings->spi_ch, REGSYNCVALUE1 + i) != 0x55) {
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: remove space after if

File: libs/pilight/hardware/raspyrfm.c [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+		return EXIT_SUCCESS;
+	} else {
+		logprintf(LOG_ERR, "the RaspyRfm module is not supported on this hardware", 0);
+		return EXIT_FAILURE;
+	}
+}
+
+int raspyrfmSend(int *code, int rawlen, int repeats, rfm_settings_t *rfmsettings) {
+	int i;
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: Initialize

File: libs/pilight/hardware/raspyrfm.c [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+	int div;
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: Initialize

File: libs/pilight/hardware/433raspyrfm.c [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+	return raspyrfmHwInit(&rfmsettings);
+}
+
+static int raspyrfm433Send(int *code, int rawlen, int repeats) {
+	return raspyrfmSend(code, rawlen, repeats, &rfmsettings);
+}
+
+static int raspyrfm433Receive(void) {
+	sleep(1);
+	return EXIT_SUCCESS;
```
Comment (pilino1234) [inline libs/pilight/hardware/433raspyrfm.c]: Placeholder code?

File: libs/pilight/hardware/raspyrfm.c [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+			return EXIT_FAILURE;
+		}
+
+		if (wiringXSPISetup(rfmsettings->spi_ch, 250000) == -1) {
+			logprintf(LOG_ERR, "failed to open SPI channel: %d", rfmsettings->spi_ch);
+			return EXIT_FAILURE;
+		}
+
+		//check if rfm69 present
+		int i;
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: Initialize

File: libs/pilight/hardware/raspyrfm.c [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+		}
+
+		if (i < 8) {
+			logprintf(LOG_ERR, "RaspyRfm module not found!");
+			return EXIT_FAILURE;
+		}
+
+		return EXIT_SUCCESS;
+	} else {
+		logprintf(LOG_ERR, "the RaspyRfm module is not supported on this hardware", 0);
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: RaspyRFM or RaspyRfm?

File: libs/pilight/hardware/raspyrfm.c [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+
+	unsigned char *fifo = MALLOC(fifoByteCnt);
+	memset(fifo, 0x00, fifoByteCnt);
+
+	unsigned char mask = 0x80;
+	unsigned char *p = fifo;
+	int r = 0;
+	for (r=0; r<repeats; r++) {
+		for (i=0; i<rawlen; i++) {
+			int c;
```
Comment (pilino1234) [inline libs/pilight/hardware/raspyrfm.c]: initialize

File: libs/pilight/hardware/868raspyrfm.c [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+	return raspyrfmHwInit(&rfmsettings);
+}
+
+static int raspyrfm868Send(int *code, int rawlen, int repeats) {
+	return raspyrfmSend(code, rawlen, repeats, &rfmsettings);
+}
+
+static int raspyrfm868Receive(void) {
+	sleep(1);
+	return EXIT_SUCCESS;
```
Comment (pilino1234) [inline libs/pilight/hardware/868raspyrfm.c]: Placeholder code? Receiving not supported?

**Comment by CurlyMoo** [comment]
Are you also planning to implement receive support?

**Comment by Phunkafizer** [comment]
Yes, but I'm still thinking about the way to do. There are different options, but unfortunately every option has different disadvantages. The main problem is that the output pin for raw-data of the RaspyRFM is not set up by default.

Option 1: Simply close the solder jumper on the bottom of the RaspyRFM. Disadvantage: Not possible to usw with the twin-RaspyRFM

Option 2: Scratch the IRQ path on the board and cross connect it to the raw pin. Disadvantage: missing IRQ line, module will not work with other software anymore that needs the IRQ line (the pilight implementation doesn't need the IRQ line)

Option 3: Solder a jumperwire to the raw pin and connect it to a free GPIO on the raspberry. Disadvantage: additional pins to connect

I'm testing a further option at the moment to use the internal packet engine and getting the received data over SPI, but at the moment it seems that this will not work.

So maybe the best solution is simply to specify a RX pin as you do in the 433gpio, then the user can select one of the options.

What do you think?

S>>

**Comment by CurlyMoo** [comment]
How does the device currently handle received messages? Over SPI?

**Comment by wo-rasp** [comment]
I would prefer the same method as for 433 and let the user define the GPIO pins in the hardware section of config. json.


Von meinem iPhone gesendet

**Comment by pilino1234** [comment]
@wo-rasp There's not much choice regarding the pins on this one, see the link in the OP for the product description and pics.

**Comment by Phunkafizer** [comment]
>How does the device currently handle received messages? Over SPI?

Yes. But you have to set baudrate and syncpattern before you can receive via SPI. For the use of pilight you don't know these Parameters, you want to find them out. I was trying to receive with a high Datarate, but the internal bitsynchronizer was freaking out. So maybe the best solution is:

- specify RAW Pin in config
- Close solderjumper if only a Singleband Module
- solder 2 jumperwires if twinbandmodem

S>>

**Comment by CurlyMoo** [comment]
Can you explain this with an annotated photo?

**Comment by Phunkafizer** [comment]
Here you can see the bottom side of the RaspyRFM. For receiving raw data, SJ2 has to be connected. This will work for single modules.

For the twinmodule one could use such a cable, cut one end and solder it to the two raw output pins on top of the RaspyRFM PCB:
https://www.google.de/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&cad=rja&uact=8&ved=0ahUKEwjx1eGhsbrTAhVDDJoKHcx2DSwQjRwIBw&url=http%3A%2F%2Fwww.ebay.de%2Fitm%2F2-Pin-Dupont-Jumper-Kabel-female-weiblich-70-cm-3D-Drucker-RepRap-Arduino-%2F172033251096&psig=AFQjCNEfgVsz7ywUg-2kjJt8iLFkJEOMUw&ust=1493030245812344

**Comment by CurlyMoo** [comment]
What about making a version that just supports raw output?

**Comment by Phunkafizer** [comment]
Changing the hardwaredesign would make it incompatible to other software interfacing the RaspyRFM, and I want ist to be a universal thing.

**Comment by CurlyMoo** [comment]
> Changing the hardwaredesign would make it incompatible to other software interfacing the RaspyRFM, and I want ist to be a universal thing.

Adding a jumper to choose between modes would love that.

**Comment by Phunkafizer** [comment]
Yes, I think I'll do so for the next revision

**Comment by CurlyMoo** [comment]
What should a raw stream look like. Just interrupt durations?

**Comment by Phunkafizer** [comment]
Raw stream is the demodulated signal. If the RFM69 is configured to OOK modulation you will get nearly the same result as the 433gpio Hardware, just filterperformance and sensitivity may be better  with the RFM69.

**Comment by CurlyMoo** [comment]
Nice!

**Comment by Phunkafizer** [comment]
Yes it is, to be precise it's the RFM69CW. Meanwhile I found a good idea for the next hardwarerevision; I will simply add 2 pins to the connector, then everything including backwardcompatibility is fulfilled.

**Comment by CurlyMoo** [comment]
Just a two pin header?

**Comment by Phunkafizer** [comment]
No; at the moment it is a 10 pin header to the PI, I will extend it to 12 pin having two more GPIOs. Am 23.04.2017 9:44 nachm. schrieb CurlyMoo <notifications@github.com>:Just a two pin header?

**Comment by CurlyMoo** [comment]
Then i'll just wait buying one :smile:

**Comment by Phunkafizer** [comment]
Will take some time because I have many on stock ;-) 
You are interessted in the twin module?

**Comment by CurlyMoo** [comment]
> You are interessted in the twin module?

Do the twin modules handle both 868 and 433 (at the same time)?

**Comment by Phunkafizer** [comment]
Yes it does; maybe there is a missunderstanding: For using a singleband module for reception, connecting the jumper on bottom is enough. The additional pins are only neccessary for reception with the twinmodule.

**Comment by CurlyMoo** [comment]
Yes, i'm interested in the twin module new revision.

**Comment by CurlyMoo** [comment]
This PR contains a lot of commit not related to your code. Can you check that?

**Comment by Phunkafizer** [comment]
Hm, I had to merge because the development branches have diverged. I have to idea how to handle that now. Anyway this part is not yet finished. Maybe it is a good idea to cancel this pull request and I make a new one when it is finished. 
Before that I will make a pull request for the voltcraft support. By the way: I wrote the voltcraft support. When it is finished, can I make a pull request from my "voltcraft" branch?
https://github.com/Phunkafizer/pilight/commit/17046e33202b84851363f146bfe5a9a0b181643d
S>>

**Comment by CurlyMoo** [comment]
Just do a force push after to this branch with the code fixup'ed.

**Pull request closed without merging**